velox/exec/tests/MarkDistinctTest.cpp

```diff
@@ -0,0 +1,139 @@
+
+    auto distinctCol =
+        makeFlatVector<bool>(size, [&](auto row) { return row < baseSize; });
+
+    auto expectedResults = makeRowVector({baseEncoded, distinctCol});
+
+    auto op = PlanBuilder()
+                  .values(vectors)
+                  .markDistinct("c0$Distinct", {"c0"})
+                  .planNode();
+
+    auto results = AssertQueryBuilder(op).copyResults(pool());
+    assertEqualVectors(results, expectedResults);
+  }
+};
+
+template <typename T>
+class MarkDistinctPODTest : public MarkDistinctTest {};
+
+using MyTypes =
+    ::testing::Types<int8_t, int16_t, int32_t, int64_t, float, double, bool>;
+TYPED_TEST_SUITE(MarkDistinctPODTest, MyTypes);
+
+TYPED_TEST(MarkDistinctPODTest, basic) {
+  auto data = VectorTestBase::makeFlatVector<TypeParam>(
+      2, [&](auto row) { return row % 2; });
+
+  MarkDistinctTest::markDistinctTest(data);
+}
+
+TEST_F(MarkDistinctTest, array) {
+  auto base = makeArrayVector<int64_t>({
+      {1, 2, 3, 4, 5},
+      {1, 2, 3},
+  });
+  markDistinctTest(base);
+}
+
+TEST_F(MarkDistinctTest, map) {
+  auto base = makeMapVector<int8_t, int32_t>(
+      {{{1, 1}, {1, 1}, {1, 1}, {1, 1}, {1, 1}}, {{1, 1}, {1, 1}, {1, 1}}});
+  markDistinctTest(base);
+}
+
+TEST_F(MarkDistinctTest, varchar) {
+  auto base = makeFlatVector<StringView>({
+      "{1, 2, 3, 4, 5}",
+      "{1, 2, 3}",
+  });
+  markDistinctTest(base);
+}
+
+TEST_F(MarkDistinctTest, row) {
+  auto base = makeRowVector({
+      makeArrayVector<int64_t>({
+          {1, 2, 3, 4, 5},
+          {1, 2, 3},
+      }),
+      makeMapVector<int8_t, int32_t>({
+          {{1, 1}, {1, 1}, {1, 1}, {1, 1}, {1, 1}},
+          {{1, 1}, {1, 1}, {1, 1}},
+      }),
+  });
+  markDistinctTest(base);
+}
+
+// This test verifies this query:
+// WITH sample AS ( SELECT * FROM (VALUES (1,1,1), (1,1,2), (1,1,1), (2,1,2),
+// (2,2,1), (2,3,2)) AS account (c1, c2, c3) ) SELECT c1,sum(distinct
+// c2),sum(distinct c3) from sample group by c1;
+TEST_F(MarkDistinctTest, distinctAggregationTest) {
+  std::vector<RowVectorPtr> vectors;
+  // Simulate the input over 3 splits.
+  vectors.push_back(makeRowVector(
+      {makeFlatVector<int32_t>({1, 1}),
+       makeFlatVector<int32_t>({1, 1}),
+       makeFlatVector<int32_t>({1, 2})}));
+  vectors.push_back(makeRowVector(
+      {makeFlatVector<int32_t>({1, 2}),
+       makeFlatVector<int32_t>({1, 1}),
+       makeFlatVector<int32_t>({1, 2})}));
+  vectors.push_back(makeRowVector(
+      {makeFlatVector<int32_t>({2, 2}),
+       makeFlatVector<int32_t>({2, 3}),
+       makeFlatVector<int32_t>({1, 2})}));
+
+  auto op =
+      PlanBuilder()
+          .values(vectors)
+          .markDistinct("c1$Distinct", {"c0", "c1"})
+          .markDistinct("c2$Distinct", {"c0", "c2"})
+          .singleAggregation(
+              {"c0"}, {"sum(c1)", "sum(c2)"}, {"c1$Distinct", "c2$Distinct"})
+          .orderBy({"c0"}, false)
+          .planNode();
+
+  auto expected = makeRowVector(
```

Comment:
Is it possible to have a SQL query to verify results. Then it would be cleaner to verify with a DuckDB query runner.

Reply:
No, duckdb does not have equivalent of this operator. We are verifying the operator output.

Reply:
It sounds like SELECT c1,sum(distinct
// c2),sum(distinct c3) from sample group by c1... was what you are trying to compute with your plan. Would it be possible to use this query with DuckDB to verify results ? That way then you can add whatever input values you want in your data. That would make it possible to test more varieties of input data.

Reply:
I think the current way the test is written is clear enough and switching to SQL query to verify the results won't be justifiable.